### ISSUE TITLE
global constructors not called in WebAssembly build

### ISSUE THREAD

**Issue by alekzander2005** [kripken/emscripten#5391]
In attempt to migrate out quite big project from asm.js to wasm target we have a problem which looks like described in subject: global objects constructed without proper constructors called.

Long story short i made a test case which in my environment(emscripten 1.37.16, chrome 59, FF 54) reproduces that problem.

Compiling and running following code:
```
 #include <iostream>
 #include <string>

 class std_string
 {
 public:
    std_string() {std::cout << "std_string()\n";}
    std_string(const char* s): ptr(s) {std::cout << "std_string(const char* s) " << uint32_t(s) << std::endl;}
    std_string(const std_string& s): ptr(s.ptr) {std::cout << "std_string(const std_string& s) " <<  uint32_t(s.ptr) << std::endl;}
    //std_string& operator=(const std_string& s) {ptr = s.ptr;}
    const char* data() const {return ptr;}
private:
    const char* ptr = nullptr;
};

const std_string txtTestString("212121\0");
const std::string s2text("someweirdtext");

int main()
{
    std::cout << s2text << std::endl;

    std::cout << uint32_t(txtTestString.data()) << std::endl;
    std::cout << txtTestString.data() << std::endl;

    return 0;
}
```

compiled by
```
emcc main.cpp -o test.js -std=c++11 -s WASM=1 -s NO_EXIT_RUNTIME=1 -Oz
```

and opened by html:
```
<!DOCTYPE html>
<html>
<head>
	<meta http-equiv="Content-Type" content="text/html; charset=utf8">
    <title></title>
	<script src='test.js'></script>
</head>
<body>
</body>

</html>
```
we have output like
```

0
emcc
```
But if -Oz was removed from compilation command line - all works fine:
```
std_string(const char* s) 11288  test.js:187:7
pre-main prep time: 174 ms  test.js:190:7
someweirdtext  test.js:187:7
11288  test.js:187:7
212121  test.js:187:7
exit(0) implicitly called by end of main(), but noExitRuntime, so not exiting the runtime (you can use emscripten_force_exit, if you want to force a true shutdown)
```

So it is looks to me as a global constructors wasn't called(optimized out probably) and it is a bug.

Thank you for your wonderful work, by the way :)

**Comment by juj** [comment]
Agreed, this looks like a bug. Optimizing out global constructors is somewhat subtle. It does look like a compiler should not optimize away global ctors if they have side effects, but if one links a library to a main program, and the library contains any unreferenced objects (even if their ctors have side effects), those ctors can be optimized away.

The issue seems to be related to the ctor evaller optimization. As a workaround, you can use

`emcc main.cpp -o test.js -std=c++11 -s WASM=1 -s NO_EXIT_RUNTIME=1 -Oz -s EVAL_CTORS=0`

which seems to work correctly.

**Comment by kripken** [comment]
Yeah, turns out we didn't handle removing some but not all ctors. Fix in #5395.

Thanks for reporting!
